File: packages/simulation-core/Sources/SimulationCore/PitchKernelDomain.swift
import Foundation

public enum ZoneIntent: String, Codable, CaseIterable, Sendable {
    case strike
    case edge
    case chase
}

public struct PitchCall: Codable, Equatable, Sendable {
    public let pitchType: PitchType
    public let zone: PitchZone
    public let zoneIntent: ZoneIntent
    public let intensity: PitchIntensity

    public init(
        pitchType: PitchType,
        zone: PitchZone,
        zoneIntent: ZoneIntent,
        intensity: PitchIntensity
    ) {
        self.pitchType = pitchType
        self.zone = zone
        self.zoneIntent = zoneIntent
        self.intensity = intensity
    }
}

public struct BatterScoutingSnapshot: Codable, Equatable, Sendable {
    public let hotZone: PitchZone
    public let coldZone: PitchZone
    public let pitchStrength: PitchType
    public let pitchWeakness: PitchType
    public let chaseTendency: Int

    public init(
        hotZone: PitchZone,
        coldZone: PitchZone,
        pitchStrength: PitchType,
        pitchWeakness: PitchType,
        chaseTendency: Int
    ) {
        self.hotZone = hotZone
        self.coldZone = coldZone
        self.pitchStrength = pitchStrength
        self.pitchWeakness = pitchWeakness
        self.chaseTendency = chaseTendency
    }
}

public struct PlateAppearanceContext: Codable, Equatable, Sendable {
    public let plateAppearanceID: String
    public let revision: UInt64
    public let inning: Int
    public let outs: Int
    public let balls: Int
    public let strikes: Int
    public let pitchNumber: Int
    public let scoreDifferential: Int
    public let leverage: Int
    public let fatigue: Int

    public init(
        plateAppearanceID: String,
        revision: UInt64,
        inning: Int,
        outs: Int,
        balls: Int,
        strikes: Int,
        pitchNumber: Int,
        scoreDifferential: Int,
        leverage: Int,
        fatigue: Int
    ) {
        self.plateAppearanceID = plateAppearanceID
        self.revision = revision
        self.inning = inning
        self.outs = outs
        self.balls = balls
        self.strikes = strikes
        self.pitchNumber = pitchNumber
        self.scoreDifferential = scoreDifferential
        self.leverage = leverage
        self.fatigue = fatigue
    }
}

public struct PreparePitchParams: Codable, Equatable, Sendable {
    public let seed: String
    public let pitcher: PitcherSnapshot
    public let batter: BatterSnapshot
    public let scouting: BatterScoutingSnapshot
    public let context: PlateAppearanceContext
    public let rivalMemory: RivalMemorySnapshot?
    public let gameState: GameStateSnapshot?
    public let gameLog: GameLogSnapshot?

    public init(
        seed: String,
        pitcher: PitcherSnapshot,
        batter: BatterSnapshot,
        scouting: BatterScoutingSnapshot,
        context: PlateAppearanceContext
    ) {
        self.init(
            seed: seed,
            pitcher: pitcher,
            batter: batter,
            scouting: scouting,
            context: context,
            rivalMemory: nil,
            gameState: nil,
            gameLog: nil
        )
    }

    public init(
        seed: String,
        pitcher: PitcherSnapshot,
        batter: BatterSnapshot,
        scouting: BatterScoutingSnapshot,
        context: PlateAppearanceContext,
        rivalMemory: RivalMemorySnapshot?,
        gameState: GameStateSnapshot? = nil,
        gameLog: GameLogSnapshot? = nil
    ) {
        self.seed = seed
        self.pitcher = pitcher
        self.batter = batter
        self.scouting = scouting
        self.context = context
        self.rivalMemory = rivalMemory
        self.gameState = gameState
        self.gameLog = gameLog
    }
}

public struct SubmitPitchParams: Codable, Equatable, Sendable {
    public let seed: String
    public let pitcher: PitcherSnapshot
    public let batter: BatterSnapshot
    public let scouting: BatterScoutingSnapshot
    public let context: PlateAppearanceContext
    public let preparationToken: String
    public let call: PitchCall
    public let rivalMemory: RivalMemorySnapshot?
    public let gameState: GameStateSnapshot?
    public let gameLog: GameLogSnapshot?

    public init(
        seed: String,
        pitcher: PitcherSnapshot,
        batter: BatterSnapshot,
        scouting: BatterScoutingSnapshot,
        context: PlateAppearanceContext,
        preparationToken: String,
        call: PitchCall
    ) {
        self.init(
            seed: seed,
            pitcher: pitcher,
            batter: batter,
            scouting: scouting,
            context: context,
            preparationToken: preparationToken,
            call: call,
            rivalMemory: nil,
            gameState: nil,
            gameLog: nil
        )
    }

    public init(
        seed: String,
        pitcher: PitcherSnapshot,
        batter: BatterSnapshot,
        scouting: BatterScoutingSnapshot,
        context: PlateAppearanceContext,
        preparationToken: String,
        call: PitchCall,
        rivalMemory: RivalMemorySnapshot?,
        gameState: GameStateSnapshot? = nil,
        gameLog: GameLogSnapshot? = nil
    ) {
        self.seed = seed
        self.pitcher = pitcher
        self.batter = batter
        self.scouting = scouting
        self.context = context
        self.preparationToken = preparationToken
        self.call = call
        self.rivalMemory = rivalMemory
        self.gameState = gameState
        self.gameLog = gameLog
    }
}

public struct CatcherRecommendation: Codable, Equatable, Sendable {
    public let call: PitchCall
    public let confidence: Int
    public let reasonCodes: [String]

    public init(call: PitchCall, confidence: Int, reasonCodes: [String]) {
        self.call = call
        self.confidence = confidence
        self.reasonCodes = reasonCodes
    }
}

public struct CatcherRecommendationSnapshot: Codable, Equatable, Sendable {
    public let call: PitchCall
    public let confidence: Int
    public let reasonCodes: [String]
    public let shortReason: String

    public init(
        call: PitchCall,
        confidence: Int,
        reasonCodes: [String],
        shortReason: String
    ) {
        self.call = call
        self.confidence = confidence
        self.reasonCodes = reasonCodes
        self.shortReason = shortReason
    }
}

public struct PitchPreparation: Codable, Equatable, Sendable {
    public let seed: String
    public let revision: UInt64
    public let pitchNumber: Int
    public let preparationToken: String
    public let planCommitment: String
    public let primaryRecommendation: CatcherRecommendationSnapshot
    public let alternativeRecommendation: CatcherRecommendationSnapshot
    public let rivalAdaptation: RivalAdaptationSnapshot

    public init(
        seed: String,
        revision: UInt64,
        pitchNumber: Int,
        preparationToken: String,
        planCommitment: String,
        primaryRecommendation: CatcherRecommendationSnapshot,
        alternativeRecommendation: CatcherRecommendationSnapshot,
        rivalAdaptation: RivalAdaptationSnapshot
    ) {
        self.seed = seed
        self.revision = revision
        self.pitchNumber = pitchNumber
        self.preparationToken = preparationToken
        self.planCommitment = planCommitment
        self.primaryRecommendation = primaryRecommendation
        self.alternativeRecommendation = alternativeRecommendation
        self.rivalAdaptation = rivalAdaptation
    }
}

public enum SelectionQuality: String, Codable, Sendable {
    case poor
    case risky
    case good
    case excellent
}

public enum PlateAppearanceResult: String, Codable, Sendable {
    case strikeout
    case walk
    case inPlayOut = "in_play_out"
    case hit
}

public struct PitchExecution: Codable, Equatable, Sendable {
    public let targetX: Int
    public let targetY: Int
    public let actualX: Int
    public let actualY: Int
    public let velocityTenthsKPH: Int
    public let horizontalBreakTenthsCM: Int
    public let verticalBreakTenthsCM: Int
    public let executionQuality: Int
    public let flightTimeMilliseconds: Int?
    public let trajectoryControlX: Int?
    public let trajectoryControlY: Int?
    /// Flat 3D series in groups of [timeMs, lateralTenthsCM, forwardTenthsCM, heightTenthsCM].
    public let trajectorySeries: [Int]?

    public init(
        targetX: Int,
        targetY: Int,
        actualX: Int,
        actualY: Int,
        velocityTenthsKPH: Int,
        horizontalBreakTenthsCM: Int,
        verticalBreakTenthsCM: Int,
        executionQuality: Int,
        flightTimeMilliseconds: Int? = nil,
        trajectoryControlX: Int? = nil,
        trajectoryControlY: Int? = nil,
        trajectorySeries: [Int]? = nil
    ) {
        self.targetX = targetX
        self.targetY = targetY
        self.actualX = actualX
        self.actualY = actualY
        self.velocityTenthsKPH = velocityTenthsKPH
        self.horizontalBreakTenthsCM = horizontalBreakTenthsCM
        self.verticalBreakTenthsCM = verticalBreakTenthsCM
        self.executionQuality = executionQuality
        self.flightTimeMilliseconds = flightTimeMilliseconds
        self.trajectoryControlX = trajectoryControlX
        self.trajectoryControlY = trajectoryControlY
        self.trajectorySeries = trajectorySeries
    }
}

public struct BattedBall: Codable, Equatable, Sendable {
    public let exitVelocityTenthsKPH: Int
    public let launchAngleTenthsDegrees: Int
    public let directionTenthsDegrees: Int
    public let contactQuality: Int

    public init(
        exitVelocityTenthsKPH: Int,
        launchAngleTenthsDegrees: Int,
        directionTenthsDegrees: Int,
        contactQuality: Int
    ) {
        self.exitVelocityTenthsKPH = exitVelocityTenthsKPH
        self.launchAngleTenthsDegrees = launchAngleTenthsDegrees
        self.directionTenthsDegrees = directionTenthsDegrees
        self.contactQuality = contactQuality
    }
}

public struct PitchKernelEvent: Codable, Equatable, Sendable {
    public let eventType: String
    public let sequence: Int
    public let planCommitment: String?
    public let primaryRecommendation: CatcherRecommendation?
    public let alternativeRecommendation: CatcherRecommendation?
    public let call: PitchCall?
    public let execution: PitchExecution?
    public let outcome: PitchOutcome?
    public let battedBall: BattedBall?
    public let fieldingResolution: FieldingResolutionSnapshot?
    public let baserunnerAdvance: BaserunnerAdvanceSnapshot?
    public let stealAttempt: StealAttemptSnapshot?
    public let inningTransition: InningTransitionSnapshot?
    public let plateAppearanceResult: PlateAppearanceResult?
    public let rivalAdaptation: RivalAdaptationSnapshot?
    public let postgameAnalysis: PostgameAnalysisSnapshot?
    public let reasonCodes: [String]

    public init(
        eventType: String,
        sequence: Int,
        planCommitment: String? = nil,
        primaryRecommendation: CatcherRecommendation? = nil,
        alternativeRecommendation: CatcherRecommendation? = nil,
        call: PitchCall? = nil,
        execution: PitchExecution? = nil,
        outcome: PitchOutcome? = nil,
        battedBall: BattedBall? = nil,
        fieldingResolution: FieldingResolutionSnapshot? = nil,
        baserunnerAdvance: BaserunnerAdvanceSnapshot? = nil,
        stealAttempt: StealAttemptSnapshot? = nil,
        inningTransition: InningTransitionSnapshot? = nil,
        plateAppearanceResult: PlateAppearanceResult? = nil,
        rivalAdaptation: RivalAdaptationSnapshot? = nil,
        postgameAnalysis: PostgameAnalysisSnapshot? = nil,
        reasonCodes: [String] = []
    ) {
        self.eventType = eventType
        self.sequence = sequence
        self.planCommitment = planCommitment
        self.primaryRecommendation = primaryRecommendation
        self.alternativeRecommendation = alternativeRecommendation
        self.call = call
        self.execution = execution
        self.outcome = outcome
        self.battedBall = battedBall
        self.fieldingResolution = fieldingResolution
        self.baserunnerAdvance = baserunnerAdvance
        self.stealAttempt = stealAttempt
        self.inningTransition = inningTransition
        self.plateAppearanceResult = plateAppearanceResult
        self.rivalAdaptation = rivalAdaptation
        self.postgameAnalysis = postgameAnalysis
        self.reasonCodes = reasonCodes
    }
}

public struct PlateAppearanceSnapshot: Codable, Equatable, Sendable {
    public let revision: UInt64
    public let balls: Int
    public let strikes: Int
    public let pitchNumber: Int
    public let ended: Bool
    public let result: PlateAppearanceResult?
    public let outcome: PitchOutcome
    public let selectionQuality: SelectionQuality
    public let recommendationAccepted: Bool
    public let fatigueAfterPitch: Int
    public let execution: PitchExecution
    public let battedBall: BattedBall?
    public let fieldingResolution: FieldingResolutionSnapshot?
    public let runnersBefore: BaserunnerStateSnapshot?
    public let runnersAfter: BaserunnerStateSnapshot
    public let runsScored: Int
    public let stealAttempt: StealAttemptSnapshot?
    public let inningTransition: InningTransitionSnapshot?
    public let reasonCodes: [String]
    public let shortFeedback: String
    public let detailFeedback: String
    public let accessibilitySummary: String

    public init(
        revision: UInt64,
        balls: Int,
        strikes: Int,
        pitchNumber: Int,
        ended: Bool,
        result: PlateAppearanceResult?,
        outcome: PitchOutcome,
        selectionQuality: SelectionQuality,
        recommendationAccepted: Bool,
        fatigueAfterPitch: Int,
        execution: PitchExecution,
        battedBall: BattedBall?,
        fieldingResolution: FieldingResolutionSnapshot? = nil,
        runnersBefore: BaserunnerStateSnapshot? = nil,
        runnersAfter: BaserunnerStateSnapshot = .empty,
        runsScored: Int = 0,
        stealAttempt: StealAttemptSnapshot? = nil,
        inningTransition: InningTransitionSnapshot? = nil,
        reasonCodes: [String],
        shortFeedback: String,
        detailFeedback: String,
        accessibilitySummary: String
    ) {
        self.revision = revision
        self.balls = balls
        self.strikes = strikes
        self.pitchNumber = pitchNumber
        self.ended = ended
        self.result = result
        self.outcome = outcome
        self.selectionQuality = selectionQuality
        self.recommendationAccepted = recommendationAccepted
        self.fatigueAfterPitch = fatigueAfterPitch
        self.execution = execution
        self.battedBall = battedBall
        self.fieldingResolution = fieldingResolution
        self.runnersBefore = runnersBefore
        self.runnersAfter = runnersAfter
        self.runsScored = runsScored
        self.stealAttempt = stealAttempt
        self.inningTransition = inningTransition
        self.reasonCodes = reasonCodes
        self.shortFeedback = shortFeedback
        self.detailFeedback = detailFeedback
        self.accessibilitySummary = accessibilitySummary
    }
}

public struct PitchKernelResult: Codable, Equatable, Sendable {
    public let revision: UInt64
    public let nextSeed: String
    public let events: [PitchKernelEvent]
    public let snapshot: PlateAppearanceSnapshot
    public let nextPreparation: PitchPreparation?
    public let rivalMemory: RivalMemorySnapshot
    public let rivalAdaptation: RivalAdaptationSnapshot
    public let gameState: GameStateSnapshot
    public let gameLog: GameLogSnapshot
    public let postgameAnalysis: PostgameAnalysisSnapshot
    public let eventHash: String

    public init(
        revision: UInt64,
        nextSeed: String,
        events: [PitchKernelEvent],
        snapshot: PlateAppearanceSnapshot,
        nextPreparation: PitchPreparation?,
        rivalMemory: RivalMemorySnapshot,
        rivalAdaptation: RivalAdaptationSnapshot,
        gameState: GameStateSnapshot,
        gameLog: GameLogSnapshot,
        postgameAnalysis: PostgameAnalysisSnapshot,
        eventHash: String
    ) {
        self.revision = revision
        self.nextSeed = nextSeed
        self.events = events
        self.snapshot = snapshot
        self.nextPreparation = nextPreparation
        self.rivalMemory = rivalMemory
        self.rivalAdaptation = rivalAdaptation
        self.gameState = gameState
        self.gameLog = gameLog
        self.postgameAnalysis = postgameAnalysis
        self.eventHash = eventHash
    }
}

File: apps/android-unity/Assets/Game/Core/Pitching/PitchKernelDomain.cs
using System;
using System.Collections.Generic;
using Baseball.Core.Domain;

namespace Baseball.Core.Pitching
{
    public enum ZoneIntent { Strike, Edge, Chase }
    public enum SelectionQuality { Poor, Risky, Good, Excellent }
    public enum PlateAppearanceResult { Strikeout, Walk, InPlayOut, Hit }

    public static class PitchingWire
    {
        public static string Value(this ZoneIntent value)
        {
            switch (value)
            {
                case ZoneIntent.Strike: return "strike";
                case ZoneIntent.Edge: return "edge";
                case ZoneIntent.Chase: return "chase";
                default: throw new ArgumentOutOfRangeException(nameof(value));
            }
        }

        public static string Value(this SelectionQuality value)
        {
            switch (value)
            {
                case SelectionQuality.Poor: return "poor";
                case SelectionQuality.Risky: return "risky";
                case SelectionQuality.Good: return "good";
                case SelectionQuality.Excellent: return "excellent";
                default: throw new ArgumentOutOfRangeException(nameof(value));
            }
        }

        public static string Value(this PlateAppearanceResult value)
        {
            switch (value)
            {
                case PlateAppearanceResult.Strikeout: return "strikeout";
                case PlateAppearanceResult.Walk: return "walk";
                case PlateAppearanceResult.InPlayOut: return "in_play_out";
                case PlateAppearanceResult.Hit: return "hit";
                default: throw new ArgumentOutOfRangeException(nameof(value));
            }
        }
    }

    public static class ZoneIntentRules
    {
        public static IReadOnlyList<ZoneIntent> Options(PitchZone zone)
        {
            return zone.Row == 1 && zone.Column == 1
                ? new[] { ZoneIntent.Strike, ZoneIntent.Chase }
                : new[] { ZoneIntent.Strike, ZoneIntent.Edge, ZoneIntent.Chase };
        }

        public static ZoneIntent Clamp(ZoneIntent intent, PitchZone zone)
        {
            return zone.Row == 1 && zone.Column == 1 && intent == ZoneIntent.Edge
                ? ZoneIntent.Strike
                : intent;
        }
    }

    public sealed class PitchCall : IEquatable<PitchCall>
    {
        public PitchCall(PitchType pitchType, PitchZone zone, ZoneIntent zoneIntent, PitchIntensity intensity)
        {
            PitchType = pitchType;
            Zone = zone;
            ZoneIntent = zoneIntent;
            Intensity = intensity;
        }

        public PitchType PitchType { get; }
        public PitchZone Zone { get; }
        public ZoneIntent ZoneIntent { get; }
        public PitchIntensity Intensity { get; }

        public bool Equals(PitchCall other)
        {
            return other != null && PitchType == other.PitchType && Zone == other.Zone &&
                   ZoneIntent == other.ZoneIntent && Intensity == other.Intensity;
        }

        public override bool Equals(object obj) => Equals(obj as PitchCall);
        public override int GetHashCode() => unchecked((((int)PitchType * 397) ^ Zone.GetHashCode()) * 397 ^ (int)ZoneIntent);
    }

    public sealed class BatterScoutingSnapshot
    {
        public BatterScoutingSnapshot(
            PitchZone hotZone,
            PitchZone coldZone,
            PitchType pitchStrength,
            PitchType pitchWeakness,
            int chaseTendency,
            int reliability = ScoutingEstimate.TrustedReliability)
        {
            HotZone = hotZone;
            ColdZone = coldZone;
            PitchStrength = pitchStrength;
            PitchWeakness = pitchWeakness;
            ChaseTendency = chaseTendency;
            Reliability = reliability;
        }

        public PitchZone HotZone { get; }
        public PitchZone ColdZone { get; }
        public PitchType PitchStrength { get; }
        public PitchType PitchWeakness { get; }
        public int ChaseTendency { get; }
        public int Reliability { get; }
    }

    public sealed class ScoutingReportSnapshot
    {
        public ScoutingReportSnapshot(
            int reliability,
            int observationCount,
            string band,
            PitchType estimatedWeakness,
            PitchZone estimatedColdZone,
            PitchType? estimatedStrength,
            PitchZone? estimatedHotZone,
            int estimatedChaseTendency,
            int chaseTendencyMargin)
        {
            Reliability = reliability;
            ObservationCount = observationCount;
            Band = band;
            EstimatedWeakness = estimatedWeakness;
            EstimatedColdZone = estimatedColdZone;
            EstimatedStrength = estimatedStrength;
            EstimatedHotZone = estimatedHotZone;
            EstimatedChaseTendency = estimatedChaseTendency;
            ChaseTendencyMargin = chaseTendencyMargin;
        }

        public int Reliability { get; }
        public int ObservationCount { get; }
        public string Band { get; }
        public PitchType EstimatedWeakness { get; }
        public PitchZone EstimatedColdZone { get; }
        public PitchType? EstimatedStrength { get; }
        public PitchZone? EstimatedHotZone { get; }
        public int EstimatedChaseTendency { get; }
        public int ChaseTendencyMargin { get; }
    }

    public sealed class PlateAppearanceContext
    {
        public PlateAppearanceContext(
            string plateAppearanceId,
            ulong revision,
            int inning,
            int outs,
            int balls,
            int strikes,
            int pitchNumber,
            int scoreDifferential,
            int leverage,
            int fatigue)
        {
            PlateAppearanceId = plateAppearanceId;
            Revision = revision;
            Inning = inning;
            Outs = outs;
            Balls = balls;
            Strikes = strikes;
            PitchNumber = pitchNumber;
            ScoreDifferential = scoreDifferential;
            Leverage = leverage;
            Fatigue = fatigue;
        }

        public string PlateAppearanceId { get; }
        public ulong Revision { get; }
        public int Inning { get; }
        public int Outs { get; }
        public int Balls { get; }
        public int Strikes { get; }
        public int PitchNumber { get; }
        public int ScoreDifferential { get; }
        public int Leverage { get; }
        public int Fatigue { get; }
    }

    public sealed class PreparePitchParams
    {
        public PreparePitchParams(
            string seed,
            PitcherSnapshot pitcher,
            BatterSnapshot batter,
            BatterScoutingSnapshot scouting,
            PlateAppearanceContext context,
            RivalMemorySnapshot rivalMemory = null,
            GameStateSnapshot gameState = null,
            GameLogSnapshot gameLog = null)
        {
            Seed = seed;
            Pitcher = pitcher;
            Batter = batter;
            Scouting = scouting;
            Context = context;
            RivalMemory = rivalMemory;
            GameState = gameState;
            GameLog = gameLog;
        }

        public string Seed { get; }
        public PitcherSnapshot Pitcher { get; }
        public BatterSnapshot Batter { get; }
        public BatterScoutingSnapshot Scouting { get; }
        public PlateAppearanceContext Context { get; }
        public RivalMemorySnapshot RivalMemory { get; }
        public GameStateSnapshot GameState { get; }
        public GameLogSnapshot GameLog { get; }
    }

    public sealed class SubmitPitchParams
    {
        public SubmitPitchParams(
            string seed,
            PitcherSnapshot pitcher,
            BatterSnapshot batter,
            BatterScoutingSnapshot scouting,
            PlateAppearanceContext context,
            string preparationToken,
            PitchCall call,
            RivalMemorySnapshot rivalMemory = null,
            GameStateSnapshot gameState = null,
            GameLogSnapshot gameLog = null,
            PersonalityTrait? trait = null)
        {
            Seed = seed;
            Pitcher = pitcher;
            Batter = batter;
            Scouting = scouting;
            Context = context;
            PreparationToken = preparationToken;
            Call = call;
            RivalMemory = rivalMemory;
            GameState = gameState;
            GameLog = gameLog;
            Trait = trait;
        }

        public string Seed { get; }
        public PitcherSnapshot Pitcher { get; }
        public BatterSnapshot Batter { get; }
        public BatterScoutingSnapshot Scouting { get; }
        public PlateAppearanceContext Context { get; }
        public string PreparationToken { get; }
        public PitchCall Call { get; }
        public RivalMemorySnapshot RivalMemory { get; }
        public GameStateSnapshot GameState { get; }
        public GameLogSnapshot GameLog { get; }
        public PersonalityTrait? Trait { get; }
    }

    public sealed class CatcherRecommendation
    {
        public CatcherRecommendation(PitchCall call, int confidence, IReadOnlyList<string> reasonCodes)
        {
            Call = call;
            Confidence = confidence;
            ReasonCodes = reasonCodes;
        }

        public PitchCall Call { get; }
        public int Confidence { get; }
        public IReadOnlyList<string> ReasonCodes { get; }
    }

    public sealed class CatcherRecommendationSnapshot
    {
        public CatcherRecommendationSnapshot(PitchCall call, int confidence, IReadOnlyList<string> reasonCodes, string shortReason)
        {
            Call = call;
            Confidence = confidence;
            ReasonCodes = reasonCodes;
            ShortReason = shortReason;
        }

        public PitchCall Call { get; }
        public int Confidence { get; }
        public IReadOnlyList<string> ReasonCodes { get; }
        public string ShortReason { get; }
    }

    public sealed class PitchPreparation
    {
        public PitchPreparation(
            string seed,
            ulong revision,
            int pitchNumber,
            string preparationToken,
            string planCommitment,
            CatcherRecommendationSnapshot primaryRecommendation,
            CatcherRecommendationSnapshot alternativeRecommendation,
            RivalAdaptationSnapshot rivalAdaptation,
            ScoutingReportSnapshot scoutingReport)
        {
            Seed = seed;
            Revision = revision;
            PitchNumber = pitchNumber;
            PreparationToken = preparationToken;
            PlanCommitment = planCommitment;
            PrimaryRecommendation = primaryRecommendation;
            AlternativeRecommendation = alternativeRecommendation;
            RivalAdaptation = rivalAdaptation;
            ScoutingReport = scoutingReport;
        }

        public string Seed { get; }
        public ulong Revision { get; }
        public int PitchNumber { get; }
        public string PreparationToken { get; }
        public string PlanCommitment { get; }
        public CatcherRecommendationSnapshot PrimaryRecommendation { get; }
        public CatcherRecommendationSnapshot AlternativeRecommendation { get; }
        public RivalAdaptationSnapshot RivalAdaptation { get; }
        public ScoutingReportSnapshot ScoutingReport { get; }
    }

    public sealed class PitchExecution
    {
        public PitchExecution(
            int targetX, int targetY, int actualX, int actualY, int velocityTenthsKph,
            int horizontalBreakTenthsCm, int verticalBreakTenthsCm, int executionQuality,
            int? flightTimeMilliseconds = null, int? trajectoryControlX = null,
            int? trajectoryControlY = null, IReadOnlyList<int> trajectorySeries = null)
        {
            TargetX = targetX;
            TargetY = targetY;
            ActualX = actualX;
            ActualY = actualY;
            VelocityTenthsKph = velocityTenthsKph;
            HorizontalBreakTenthsCm = horizontalBreakTenthsCm;
            VerticalBreakTenthsCm = verticalBreakTenthsCm;
            ExecutionQuality = executionQuality;
            FlightTimeMilliseconds = flightTimeMilliseconds;
            TrajectoryControlX = trajectoryControlX;
            TrajectoryControlY = trajectoryControlY;
            TrajectorySeries = trajectorySeries;
        }

        public int TargetX { get; }
        public int TargetY { get; }
        public int ActualX { get; }
        public int ActualY { get; }
        public int VelocityTenthsKph { get; }
        public int HorizontalBreakTenthsCm { get; }
        public int VerticalBreakTenthsCm { get; }
        public int ExecutionQuality { get; }
        public int? FlightTimeMilliseconds { get; }
        public int? TrajectoryControlX { get; }
        public int? TrajectoryControlY { get; }
        public IReadOnlyList<int> TrajectorySeries { get; }
    }

    public sealed class BattedBall
    {
        public BattedBall(int exitVelocityTenthsKph, int launchAngleTenthsDegrees, int directionTenthsDegrees, int contactQuality)
        {
            ExitVelocityTenthsKph = exitVelocityTenthsKph;
            LaunchAngleTenthsDegrees = launchAngleTenthsDegrees;
            DirectionTenthsDegrees = directionTenthsDegrees;
            ContactQuality = contactQuality;
        }

        public int ExitVelocityTenthsKph { get; }
        public int LaunchAngleTenthsDegrees { get; }
        public int DirectionTenthsDegrees { get; }
        public int ContactQuality { get; }
    }

    public sealed class PitchKernelEvent
    {
        public PitchKernelEvent(
            string eventType,
            int sequence,
            string planCommitment = null,
            CatcherRecommendation primaryRecommendation = null,
            CatcherRecommendation alternativeRecommendation = null,
            PitchCall call = null,
            PitchExecution execution = null,
            PitchOutcome? outcome = null,
            BattedBall battedBall = null,
            FieldingResolutionSnapshot fieldingResolution = null,
            BaserunnerAdvanceSnapshot baserunnerAdvance = null,
            StealAttemptSnapshot stealAttempt = null,
            InningTransitionSnapshot inningTransition = null,
            PlateAppearanceResult? plateAppearanceResult = null,
            RivalAdaptationSnapshot rivalAdaptation = null,
            PostgameAnalysisSnapshot postgameAnalysis = null,
            IReadOnlyList<string> reasonCodes = null)
        {
            EventType = eventType;
            Sequence = sequence;
            PlanCommitment = planCommitment;
            PrimaryRecommendation = primaryRecommendation;
            AlternativeRecommendation = alternativeRecommendation;
            Call = call;
            Execution = execution;
            Outcome = outcome;
            BattedBall = battedBall;
            FieldingResolution = fieldingResolution;
            BaserunnerAdvance = baserunnerAdvance;
            StealAttempt = stealAttempt;
            InningTransition = inningTransition;
            PlateAppearanceResult = plateAppearanceResult;
            RivalAdaptation = rivalAdaptation;
            PostgameAnalysis = postgameAnalysis;
            ReasonCodes = reasonCodes ?? Array.Empty<string>();
        }

        public string EventType { get; }
        public int Sequence { get; }
        public string PlanCommitment { get; }
        public CatcherRecommendation PrimaryRecommendation { get; }
        public CatcherRecommendation AlternativeRecommendation { get; }
        public PitchCall Call { get; }
        public PitchExecution Execution { get; }
        public PitchOutcome? Outcome { get; }
        public BattedBall BattedBall { get; }
        public FieldingResolutionSnapshot FieldingResolution { get; }
        public BaserunnerAdvanceSnapshot BaserunnerAdvance { get; }
        public StealAttemptSnapshot StealAttempt { get; }
        public InningTransitionSnapshot InningTransition { get; }
        public PlateAppearanceResult? PlateAppearanceResult { get; }
        public RivalAdaptationSnapshot RivalAdaptation { get; }
        public PostgameAnalysisSnapshot PostgameAnalysis { get; }
        public IReadOnlyList<string> ReasonCodes { get; }
    }

    public sealed class PlateAppearanceSnapshot
    {
        public PlateAppearanceSnapshot(
            ulong revision, int balls, int strikes, int pitchNumber, bool ended,
            PlateAppearanceResult? result, PitchOutcome outcome, SelectionQuality selectionQuality,
            bool recommendationAccepted, int fatigueAfterPitch, PitchExecution execution,
            BattedBall battedBall, FieldingResolutionSnapshot fieldingResolution,
            BaserunnerStateSnapshot runnersBefore, BaserunnerStateSnapshot runnersAfter,
            int runsScored, StealAttemptSnapshot stealAttempt, InningTransitionSnapshot inningTransition,
            IReadOnlyList<string> reasonCodes, string shortFeedback, string detailFeedback,
            string accessibilitySummary)
        {
            Revision = revision;
            Balls = balls;
            Strikes = strikes;
            PitchNumber = pitchNumber;
            Ended = ended;
            Result = result;
            Outcome = outcome;
            SelectionQuality = selectionQuality;
            RecommendationAccepted = recommendationAccepted;
            FatigueAfterPitch = fatigueAfterPitch;
            Execution = execution;
            BattedBall = battedBall;
            FieldingResolution = fieldingResolution;
            RunnersBefore = runnersBefore;
            RunnersAfter = runnersAfter;
            RunsScored = runsScored;
            StealAttempt = stealAttempt;
            InningTransition = inningTransition;
            ReasonCodes = reasonCodes;
            ShortFeedback = shortFeedback;
            DetailFeedback = detailFeedback;
            AccessibilitySummary = accessibilitySummary;
        }

        public ulong Revision { get; }
        public int Balls { get; }
        public int Strikes { get; }
        public int PitchNumber { get; }
        public bool Ended { get; }
        public PlateAppearanceResult? Result { get; }
        public PitchOutcome Outcome { get; }
        public SelectionQuality SelectionQuality { get; }
        public bool RecommendationAccepted { get; }
        public int FatigueAfterPitch { get; }
        public PitchExecution Execution { get; }
        public BattedBall BattedBall { get; }
        public FieldingResolutionSnapshot FieldingResolution { get; }
        public BaserunnerStateSnapshot RunnersBefore { get; }
        public BaserunnerStateSnapshot RunnersAfter { get; }
        public int RunsScored { get; }
        public StealAttemptSnapshot StealAttempt { get; }
        public InningTransitionSnapshot InningTransition { get; }
        public IReadOnlyList<string> ReasonCodes { get; }
        public string ShortFeedback { get; }
        public string DetailFeedback { get; }
        public string AccessibilitySummary { get; }
    }

    public sealed class PitchKernelResult
    {
        public PitchKernelResult(
            ulong revision,
            string nextSeed,
            PlateAppearanceSnapshot snapshot,
            PitchPreparation nextPreparation,
            RivalMemorySnapshot rivalMemory,
            RivalAdaptationSnapshot rivalAdaptation,
            GameStateSnapshot gameState,
            GameLogSnapshot gameLog,
            PostgameAnalysisSnapshot postgameAnalysis,
            string eventHash,
            IReadOnlyList<PitchKernelEvent> events)
        {
            Revision = revision;
            NextSeed = nextSeed;
            Snapshot = snapshot;
            NextPreparation = nextPreparation;
            RivalMemory = rivalMemory;
            RivalAdaptation = rivalAdaptation;
            GameState = gameState;
            GameLog = gameLog;
            PostgameAnalysis = postgameAnalysis;
            EventHash = eventHash;
            Events = events;
        }

        public ulong Revision { get; }
        public string NextSeed { get; }
        public PlateAppearanceSnapshot Snapshot { get; }
        public PitchPreparation NextPreparation { get; }
        public RivalMemorySnapshot RivalMemory { get; }
        public RivalAdaptationSnapshot RivalAdaptation { get; }
        public GameStateSnapshot GameState { get; }
        public GameLogSnapshot GameLog { get; }
        public PostgameAnalysisSnapshot PostgameAnalysis { get; }
        public string EventHash { get; }
        public IReadOnlyList<PitchKernelEvent> Events { get; }
        public IReadOnlyList<string> EventTypes
        {
            get
            {
                var values = new string[Events.Count];
                for (var index = 0; index < Events.Count; index++) values[index] = Events[index].EventType;
                return values;
            }
        }
    }
}
